### ISSUE TITLE
Can't use Session to store access token

### ISSUE THREAD

**Issue by MartinDawson** [octokit/octokit.net#1883]
Asp.net core session is always null when I call my server through GitHub.com from my google extension.

If I do the exact same through my own localhost:8080 website then it works fine but it seems that requests from external domains don't persist the cookie for the session.

Startup.cs

```
services.AddSession(options => new SessionOptions
{
    Cookie = new CookieBuilder
    {
        SameSite = SameSiteMode.None,
        IsEssential = true,
        Name = ".onboardor",
    }
});

app.UseSession(new SessionOptions
{
    Cookie = new CookieBuilder
    {
        SameSite = SameSiteMode.None,
        IsEssential = true,
        Name = ".onboardor",
    }
});
```

No idea why their are two ways to pass sessionoptions...

```
public IActionResult Test()
{
    var value = HttpContext.Session.GetString("Test");

    HttpContext.Session.SetString("Test", "randomValue");

    return null;
}
js:

fetch("https://localhost:8080/test", {
  credentials: "include",
  headers: {
    "Accept": "application/json",
    "Content-Type": "application/json",
  },
  method: "get",
});
```

Because I can't save the access token to the session I can't authenticate my OAUTH app for subsequent requests...

**Comment by ryangribble** [comment]
Hey @MartinDawson I'm not entirely sure what you're getting at sorry, it doesn't seem to be related to octokit.net

Your comment about storing an access token in a cookie/session doesn't sound correct though - if you have human interaction you should use the oauth flow, and if you are a bot type system then you could store an access token in the app settings rather than a session, or better yet use a github app for more granular control on things 

Have you read the info on the github developer site about authentication options? 

https://developer.github.com/v3/#authentication

If you do have an issue related to this library (octokit.net) can you please clarify a bit more? 

Thanks

**Comment by ryangribble** [comment]
Ah sorry I thought you were referring to a personal access token 

So it sounds like the fact there's a chrome extension involved is a complicating factor?  I'm still not quite putting everything together here though sorry - are you developing a chrome extension yourself and trying to have it perform the oauth flow (I can't see how that works since you need github to hit a url to perform some of the flow) or are you saying you have a chrome extension installed that is somehow affecting what you're trying to do with an asp.net website that is trying to do oauth flow with github?

**Comment by MartinDawson** [comment]
Yeah, thanks for the help.

Here's the full flow:

I am creating a chrome extension for GitHub, similar to ZenHub extension. This chrome extension inject a button on the GitHub page. See screenshot below.

![screenshot](https://user-images.githubusercontent.com/15030491/46465132-3c120280-c7c8-11e8-8797-5277d8e2963f.png)

This button then calls my /setup route on my server:
I'm using GraphQL for my backend and not rest API but it shouldn't make a difference.

```c#
public override object MutateAndGetPayload(MutationInputs inputs, ResolveFieldContext<object> context)
{
    var c = context.UserContext.As<Context>();
    var csrf = Password.Generate(24, 1);

    c.HttpContext.Session.SetString("CSRF", csrf);

    var request = new Octokit.OauthLoginRequest(Env.GetString("CLIENT_ID"))
    {
        Scopes = { "repo" },
        State = csrf,
    };
    var redirectUri = inputs.Get<string>("redirectUrl");

    _logger.Log(LogLevel.Debug, $"csrf set: {csrf}");

    if (redirectUri != null)
    {
        request.RedirectUri = new Uri($"{Env.GetString("APP_URL")}/setupCallback?redirectUrl={redirectUri}");
    }

    var oAuthLoginUrl = _client.Oauth.GetGitHubLoginUrl(request);

    return new {
        oAuthLoginUrl
    };
}
```

I then redirect to that redirectUrl on my client side immediately.

In the /setupCallback I do this:

```c#
Field<NonNullGraphType<BooleanGraphType>>()
    .Argument<NonNullGraphType<StringGraphType>>("code", "The code for the setup")
    .Argument<NonNullGraphType<StringGraphType>>("state", "The CSRF protection state")
    .Name("setupCallback")
    .ResolveAsync(async c =>
    {
        var context = c.UserContext.As<Context>();
        var expectedState = context.HttpContext.Session.GetString("CSRF");
        var state = c.GetArgument<string>("state");
        var code = c.GetArgument<string>("code");

        if (state != expectedState) {
            logger.Log(LogLevel.Error, $"state: ${state} does not match code: ${code}");

            throw new InvalidOperationException("Security fail");
        }

        context.HttpContext.Session.SetString("CSRF", "");
        var request = new Octokit.OauthTokenRequest(Env.GetString("CLIENT_ID"), Env.GetString("CLIENT_SECRET"), code);
        var token = await _client.Oauth.CreateAccessToken(request);
        _client.Credentials = new Octokit.Credentials(token.AccessToken);
  
        context.HttpContext.Session.SetString("GitHubToken", token.AccessToken);

        return true;
    });
```

So basically the user will be on GitHub.com but hitting my server through my chrome extension.

The problem is that asp.net core session GitHubToken is always empty on subsequent requests. When I check the client cookies on GitHub.com my .AspnetCoreSession isn't actually there and seems to be getting removed for some reason.

This is only a problem when trying to do this through my chrome extension.

**Comment by MartinDawson** [comment]
Figured it out. Session wasn't working properly because for CORS you need to set samesite mode for cookies to none.

```c#
            app.UseCookiePolicy(new CookiePolicyOptions
            {
                MinimumSameSitePolicy = SameSiteMode.None
            });

            app.UseSession(new SessionOptions {
                Cookie = new CookieBuilder
                {
                    Name = ".AspNetCore.Session",
                    SameSite = SameSiteMode.None,
                }
            });
```
